File: isupport.go
package seabird

import (
	"strings"

	irc "gopkg.in/irc.v3"
)

func (s *Server) handleISupport(msg *irc.Message) {
	if len(msg.Params) < 2 {
		return
	}

	// Check for really old servers (or servers which based 005 off of rfc2812.
	if !strings.HasSuffix(msg.Trailing(), "server") {
		// This server doesn't appear to support ISupport messages. Here there
		// be dragons.
		return
	}

	s.isupportLock.Lock()
	defer s.isupportLock.Unlock()

	for _, param := range msg.Params[1 : len(msg.Params)-1] {
		data := strings.SplitN(param, "=", 2)
		if len(data) < 2 {
			s.isupport[data[0]] = ""
			continue
		}

		s.isupport[data[0]] = data[1]
	}
}

// IsEnabled will check for boolean ISupport values
func (s *Server) isupportIsEnabled(key string) bool {
	s.isupportLock.RLock()
	defer s.isupportLock.RUnlock()

	_, ok := s.isupport[key]
	return ok
}

// GetList will check for list ISupportValues
func (s *Server) isupportList(key string) ([]string, bool) {
	s.isupportLock.RLock()
	defer s.isupportLock.RUnlock()

	data, ok := s.isupport[key]
	if !ok {
		return nil, false
	}

	return strings.Split(data, ","), true
}

// GetMap will check for map ISupport values
func (s *Server) isupportMap(key string) (map[string]string, bool) {
	s.isupportLock.RLock()
	defer s.isupportLock.RUnlock()

	data, ok := s.isupport[key]
	if !ok {
		return nil, false
	}

	ret := make(map[string]string)

	for _, v := range strings.Split(data, ",") {
		innerData := strings.SplitN(v, ":", 2)
		if len(innerData) != 2 {
			return nil, false
		}

		ret[innerData[0]] = innerData[1]
	}

	return ret, true
}

// GetRaw will get the raw ISupport values
func (s *Server) isupportRaw(key string) (string, bool) {
	s.isupportLock.RLock()
	defer s.isupportLock.RUnlock()

	ret, ok := s.isupport[key]
	return ret, ok
}

func (s *Server) isupportPrefixMap() (map[rune]rune, bool) {
	// Sample: (qaohv)~&@%+
	prefix, _ := s.isupportRaw("PREFIX")

	// We only care about the symbols
	i := strings.IndexByte(prefix, ')')
	if len(prefix) == 0 || prefix[0] != '(' || i < 0 {
		// "Invalid prefix format"
		return nil, false
	}

	// We loop through the string using range so we get bytes, then we throw the
	// two results together in the map.
	var symbols []rune // ~&@%+
	for _, r := range prefix[i+1:] {
		symbols = append(symbols, r)
	}

	var modes []rune // qaohv
	for _, r := range prefix[1:i] {
		modes = append(modes, r)
	}

	if len(modes) != len(symbols) {
		// "Mismatched modes and symbols"
		return nil, false
	}

	prefixes := make(map[rune]rune)
	for k := range symbols {
		prefixes[symbols[k]] = modes[k]
	}

	return prefixes, true
}

File: src/irc/isupport.rs
use std::iter::FromIterator;

use crate::prelude::*;

#[derive(Debug)]
pub struct ISupportTracker {
    data: BTreeMap<String, String>,
}

impl ISupportTracker {
    pub fn new() -> Self {
        let mut data = BTreeMap::new();
        data.insert("PREFIX".to_string(), "(ov)@+".to_string());

        ISupportTracker { data: data }
    }

    pub fn handle_message(&mut self, msg: &irc::Message) {
        if msg.command != "005" {
            return;
        }

        if msg.params.len() < 2 {
            // Malformed ISupport message
            return;
        }

        // Check for really old servers (or servers which based 005 off of rfc2812.
        if !msg.params[1].ends_with("server") {
            // This server doesn't appear to support ISupport messages. Here there be dragons.
            return;
        }

        for param in msg.params[1..].as_ref() {
            let mut data = param.splitn(2, '=');
            match (data.next(), data.next()) {
                (Some(key), Some(val)) => {
                    self.data.insert(key.to_string(), val.to_string());
                }
                (Some(key), None) => {
                    self.data.insert(key.to_string(), "".to_string());
                }
                (None, _) => {
                    unreachable!();
                }
            }
        }
    }

    #[allow(dead_code)]
    pub fn get(&self, key: &str) -> Option<String> {
        self.data.get(key).map(|val| val.clone())
    }

    #[allow(dead_code)]
    pub fn is_enabled(&self, key: &str) -> bool {
        self.data.contains_key(key)
    }

    #[allow(dead_code)]
    pub fn get_list(&self, key: &str) -> Option<Vec<String>> {
        self.data
            .get(key)
            .map(|val| val.split(',').map(String::from).collect())
    }

    #[allow(dead_code)]
    pub fn get_map(&self, key: &str) -> Option<BTreeMap<String, String>> {
        self.data.get(key).and_then(|val| {
            let inner = val.split(',');

            FromIterator::from_iter(inner.map(|inner| {
                let mut split = inner.splitn(2, ':');
                match (split.next(), split.next()) {
                    (Some(k), Some(v)) => Some((k.to_string(), v.to_string())),
                    _ => None,
                }
            }))
        })
    }

    pub fn get_prefix_map(&self) -> Option<BTreeMap<u8, u8>> {
        // Sample: (qaohv)~&@%+
        self.data.get("PREFIX").and_then(|val| {
            // We only care about the symbols
            let idx = val.find(')');
            if !val.starts_with('(') || idx.is_none() {
                return None;
            }
            let idx = idx.unwrap();

            let modes = &val[1..idx];
            let symbols = &val[idx + 1..];

            if modes.len() != symbols.len() {
                // "Mismatched modes and symbols"
                return None;
            }

            Some(BTreeMap::from_iter(symbols.bytes().zip(modes.bytes())))
        })
    }
}
